teps/0130-pipeline-level-service.md

```diff
@@ -0,0 +1,759 @@
+  - A user may also want to use the built image in these subsequent tasks without having to first push to and pull from a registry;
+  for example, if using `docker exec` or `docker-compose`
+
+Allowing multiple Tasks in a PipelineRun to use the same docker daemon allows PipelineRuns with
+docker builds to take advantage of Pipeline features (e.g. using matrix to build multiple images,
+executing a push conditionally on previous steps) while still having access to the same local
+cache of built images and avoid having to push to a registry in between.
+
+#### Other uses for this feature
+
+A few other uses for Pipeline-level services have been proposed, but Docker builds are the highest priority use case identified so far.
+
+- **Integration tests**: This feature may also be useful for integration tests that use a service to communicate with a test database.
+However, integration tests are likely limited to a single Task; i.e. it's unlikely that a user would want multiple integration testing Tasks
+to share the same test database. A test database Pipeline-level sidecar service could be limited to one Task (or a subset of Tasks) using
+[Pipelines in Pipelines](./0056-pipelines-in-pipelines.md), or this use case could be met by allowing Pipelines or TaskRuns
+to specify Task-level sidecars, as proposed in
+[TEP-0126: Allow Task sidecars to be specified in PipelineRun](https://github.com/tektoncd/community/pull/877).
+
+- **Heartbeats**: The [original feature request](https://github.com/tektoncd/pipeline/issues/2973) for Pipeline-level sidecar services was within a Pipeline
+that uses Boskos resources. The first Task acquires the resources and creates a pod to send heartbeats to Boskos;
+intermediate Tasks use the Boskos resources, and a Finally Task would release the lease on the resources and clean up the pod sending heartbeats.
+One user [suggested](https://github.com/tektoncd/pipeline/issues/2973#issuecomment-672152635) that Boskos usage should be restricted to a subset of Tasks,
+which could be done with Pipelines in Pipelines.
+
+### Goals
+
+- More flexible Docker build Pipelines
+
+### Requirements
+
+- A Docker image built during a Pipeline can be used in a subsequent Task without first pushing it to an image registry
+- Service is only accessible to the single PipelineRun that uses it
+(doing otherwise would violate [SLSA L3 build isolation requirements](https://slsa.dev/spec/v0.1/requirements#isolated))
+
+### Related features in other CI/CD systems
+
+A few other CI/CD systems have sidecar-like features.
+
+The closest feature to Pipeline-level sidecar services is the Argo Workflows [daemon containers](https://argoproj.github.io/argo-workflows/walk-through/daemon-containers/) feature.
+Daemon containers share the same lifespan as a Template, which can be analogous to a Task or a Pipeline. (Argo Workflows also has a [sidecar feature](https://argoproj.github.io/argo-workflows/walk-through/sidecars/) similar to Tekton's Task sidecars.)
```

Comment:
I'm not sure what you mean here-- Argo workflows Templates are analogous to Pipelines and Tasks, and sidecars are defined in Tasks rather than TaskRuns. Would you prefer "PipelineRun-level services" to "Pipeline-level services"? I think the service would be defined in the Pipeline so I prefer "Pipeline-level services".